### PR TITLE
Backport PR12485

### LINKED ISSUE #77: Failure on Linux Python 3.7 `toolchain` build
The Linux Python 3.7 `toolchain` build is failing on CI with a segmentation fault. Here is [the build]( https://circleci.com/gh/conda-forge/scikit-learn-feedstock/516 ) and [the log](https://github.com/conda-forge/scikit-learn-feedstock/files/2499429/Linux_Python_3.7_Toolchain_Build_Log.txt) for details. We were unable to skip the build using `skip: true` with selectors. So it is left in. Help fixing either of these would be appreciated.

Edit: Have broken out the re-rendering/selector problem as issue ( https://github.com/conda-forge/conda-smithy/issues/917 ). An example of this problem is shown in PR ( https://github.com/conda-forge/scikit-learn-feedstock/pull/78 ).

### PR DESCRIPTION
**Pull request by jschueller**
Should fix #77

### CODE REVIEW

**Comment by rth** [comment]
Thanks @jschueller !

**Comment by lesteve** [comment]
I am a bit wary about fixing things in the conda-forge package because it introduces subtle differences between whether you install from pip, conda from defaults channel or conda from conda-forge channel.

My personal preference would be to have the scikit-learn fix included in 0.20.1 and to wait for 0.20.1 to be released. Is there a particular reason why this is not a workable option? I have to say I am not following the scikit-learn development that closely these days, so I am not sure how far down the line this is likely to happen.

**Comment by jschueller** [comment]
Without this it doesnt pass the tests and the py3.7 package cannot be published.

**Comment by rth** [comment]
I'm also not so keen on applying patches on conda forge, but the issue is that currently one test segfault in one of the builds see https://github.com/conda-forge/scikit-learn-feedstock/issues/77 So either, 
 - a patch needs to be applied to skip the corresponding test (or some CLI parameter if possible)
 - or the segfault needs to be resolved

I still think a segfault warants a patch, rather than ignoring it, because it has also a potential to segfault for users.

**Comment by lesteve** [comment]
> a patch needs to be applied to skip the corresponding test (or some CLI parameter if possible)

I think this is what we have done in the past (e.g. failing tests on OSX IIRC). This would be my favourite solution by far.

> I still think a segfault warants a patch, rather than ignoring it, because it has also a potential to segfault for users.

I completely agree that this should not be ignored but I'd rather have the fix propagate uniformally through the normal release process rather than in some package-manager specific way.

I would be more than happy to hear other people's opinions too! I completely understand opinions on this topic and what is higher priority are likely to vary between people ...

**Comment by lesteve** [comment]
> I think this is what we have done in the past (e.g. failing tests on OSX IIRC). This would be my favourite solution by far.

FYI I found the PR about skipping the tests on OSX for 0.18.1 https://github.com/conda-forge/scikit-learn-feedstock/pull/40. Sorry @jschueller I have to admit I don't remember you were involved in the discussion at the time, sometimes history seems to repeat itself ;-)

**Comment by rth** [comment]
I'm fine with occasionally skipping failing tests on conda-forge.

However, a segfault is different -- it has the potential to bring down a whole application running in production. I think this outweighs small inconsistencies between source code on conda and PyPi (one line change here), particularly that the differences due to different versions of compilers, numpy versions used for build etc are inevitable anyway.

**Comment by jakirkham** [comment]
As I see it, there are 3 options:

1. Don't do anything (Python 3.7 is built and available with the new compilers).
2. Merge this PR as is.
3. Disable the segfaulting test with a patch and merge.

From the compiler migration standpoint (unless I'm misunderstanding something). The intent was to only build Python 3.7 for the new compilers. Now we haven't followed that much in practice. It's not clear to me why that is as I haven't followed all of the discussions related to it as there are many, but could be an oversight that Python 3.7 is built with the old compilers at all. In which case 1 sounds like the option most inline with the compiler migration intent. Given we have made a lot of progress on the compiler migration front, switching to those packages should be a lower barrier to entry.

If we are to publish a Python 3.7 package with the old compilers, which I'm not totally convinced we should do (though please feel free to convince me), then I agree with @rth that having a segfaulting package is no good. Completely understand your point @lesteve about the burden this can cause for upstream. IIRC past discussions along these lines revolved around numerical errors, which are concerning, but not at the same level as a segfault. Given the patch is mainly about building scikit-learn on older compilers, it is pretty justifiable for conda-forge to choose to include it since we are in the business of building things here. IDK if binary wheels have this issue, but it's probably worth checking.

TL;DR The compiler migration intended only new compilers be used for Python 3.7, which makes 1 the natural choice. If for some reason, that option doesn't work, it would be good to understand more clearly why. Assuming there are good motivations for building this with the old compilers, we can discuss other options, but IMHO we should either merge this PR as is or wait for a scikit-learn patch release. Building and distributing binaries that we know segfault isn't very nice for our users; so, should be avoided.

Thoughts @jschueller? :)

**Comment by rth** [comment]
Thanks for the thorough response @jakirkham !

> If we are to publish a Python 3.7 package with the old compilers, which I'm not totally convinced we should do

Me neither. If the old compilers build doesn't need to be uploaded, then maybe there is no problem to solve; but you did open https://github.com/conda-forge/scikit-learn-feedstock/issues/77 :) Should we close it as "won't fix" then?

Generally, I don't understand why this feedstock is also built with the old compilers (unlike other scientific python feedstocks cf https://github.com/conda-forge/scikit-learn-feedstock/pull/76#issuecomment-429896096)...

**Comment by jschueller** [comment]
@jakirkham from a feedstock maintenance point of view we shouldnt have merged the migration PR with a failing test in the first place, which kind of voids option 1. Imho option 3 is sweeping the dirt under the carpet.

Also I'm not too worried about the packaging choices we make in conda-forge, the policy to make sure upstream gets patched (but dont necessarily wait on release) is a good compromise.

**Comment by amueller** [comment]
I really want to release this weekend, so either waiting or backporting doesn't seem to be that big a deal, where I would tend towards waiting.
FYI this is a bug that's unlikely to show to users imho because it's passing a read-only memory buffer to a clustering method....

**Comment by jschueller** [comment]
I propose to merge this and not wait on sklearn: https://github.com/scikit-learn/scikit-learn/milestone/27

**Comment by lesteve** [comment]
> I don't have a strong opinion but hopefully we'll release tomorrow

Great to hear! I'd rather wait for the scikit-learn release then.

**Comment by jschueller** [comment]
sklearn update still does not succeed: #80, #81
I believe this one is still worth being merged, as I still cannot install skl for my custom installer

**Comment by jnothman** [comment]
I think there is little harm in merging this. But I would love ideas about how to resolve #80!

**Comment by jakirkham** [comment]
Thanks all for helping resolve this issue.
